// snips-sam/src/commands/login.ts
import * as process from 'process'
import * as yargs from 'yargs'
import cli from '../cli'
import { ConsoleClient } from '../session'

exports.command = 'login'
exports.desc = 'Login to the console'

exports.handler = async (_: yargs.Argv) => {

    const responses = await cli.prompt.promptQuestions(
        [
            {
                name: 'email',
                type: 'email',
                message: 'Enter email used on the console:',
                validate: value => {
                    if (value.length) return true
                    return 'Email cannot be blank'
                },
            },
            {
                name: 'password',
                type: 'password',
                message: 'Enter password used on the console:',
                validate: value => {
                    if (value.length) return true
                    return 'Password cannot be blank'
                },
            },
        ],
    )

    const email = responses['email']
    const password = responses['password']

    try {
        cli.stream.loading('Logging in')
        const client = new ConsoleClient()
        await client.login(email, password)
        cli.stream.success('Logged in')
    } catch (e) {
        cli.stream.error(e.message)
    }
    process.exit()
}
