// snips-sam/src/commands/sound-feedback.ts
import { Argv } from 'yargs'
import cli from '../cli'
import { SSHService } from '../session'

exports.command = 'sound-feedback <toggle>'
exports.desc = 'Toogle sound feedback on or off'
exports.builder = yargs => {
    return yargs
    .positional('toggle', {
        describe: 'Toggle the feedback sound',
        type: 'string',
        choices: ['on', 'off'],
        demandOption: true,
    })
}

exports.handler = async (argv: Argv) => {
    const toggle: string = JSON.parse(JSON.stringify(argv)).toggle
    const mqttToggle = toggle === 'on' ? 'On' : 'Off'

    const ssh = new SSHService()
    await ssh.connect()
    .catch(e => {
        cli.stream.error(e)
        process.exit()
    })

    await ssh.checkCommandExists('mosquitto_pub')
    .then(mosquittoClientInstalled => {
        if (mosquittoClientInstalled) return
        cli.stream.loading('Installing mosquitto-clients')
        return ssh.aptGetY(['mosquitto-clients'], _ => {})
        .then(_ => cli.stream.success())
    })
    .then(_ => ssh.run(`mosquitto_pub -h localhost -p 1883 -t "hermes/feedback/sound/toggle${mqttToggle}" -m '{"siteId": "default"}'`))
    .then(_ => cli.stream.success(`Feedback sound has been turned ${toggle.toLowerCase()}`))
    .catch(e => cli.stream.error(e))
    ssh.disconnect()

}
