// snips-sam/lib/commands/install/demo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const cli_1 = require("../../cli");
const session_1 = require("../../session");
const chalk_1 = require("chalk");
const actions_1 = require("./actions");
exports.command = 'demo';
exports.desc = 'Install a weather assistant skill demo';
exports.builder = {
    ignore_action_parameters: {
        demandOption: false,
        describe: 'Do not get prompted for action parameters',
        type: 'boolean',
        alias: 'ignore',
        default: false,
    },
};
exports.handler = (argv) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const ignore = JSON.parse(JSON.stringify(argv)).ignore_action_parameters;
    const ssh = new session_1.SSHService();
    yield ssh.connect()
        .catch(e => {
        cli_1.cli.stream.error(e);
        process.exit();
    });
    cli_1.cli.stream.loading('Installing Snips fake weather demo');
    yield ssh.aptGetY(['snips-platform-demo', session_1.SnipsServices.skillServer, session_1.SnipsServices.template], _ => { });
    cli_1.cli.stream.success();
    try {
        const skillSetup = new actions_1.SkillSetup(ssh, ignore, 'https://github.com/snipsco/snips-skill-weather-tts.git');
        yield skillSetup.run();
        yield ssh.relaunchServices().catch(_ => { });
        const assistant = yield ssh.getInstalledAssistant();
        if (assistant !== undefined) {
            cli_1.cli.stream.success(`Snips assistant is now running. Say ${chalk_1.default.blue(assistant.hotword)} to start!`);
        }
        else {
            cli_1.cli.stream.success('Snips assistant is now running');
        }
        cli_1.cli.stream.hint(`Run ${chalk_1.default.blue('sam watch')} to see the logs`);
    }
    catch (error) {
        cli_1.cli.stream.error(error);
    }
    ssh.disconnect();
    process.exit();
});
