// snips-sam/lib/commands/login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const process = require("process");
const cli_1 = require("../cli");
const session_1 = require("../session");
exports.command = 'login';
exports.desc = 'Login to the console';
exports.handler = (_) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const responses = yield cli_1.default.prompt.promptQuestions([
        {
            name: 'email',
            type: 'email',
            message: 'Enter email used on the console:',
            validate: value => {
                if (value.length)
                    return true;
                return 'Email cannot be blank';
            },
        },
        {
            name: 'password',
            type: 'password',
            message: 'Enter password used on the console:',
            validate: value => {
                if (value.length)
                    return true;
                return 'Password cannot be blank';
            },
        },
    ]);
    const email = responses['email'];
    const password = responses['password'];
    try {
        cli_1.default.stream.loading('Logging in');
        const client = new session_1.ConsoleClient();
        yield client.login(email, password);
        cli_1.default.stream.success('Logged in');
    }
    catch (e) {
        cli_1.default.stream.error(e.message);
    }
    process.exit();
});
