// snips-sam/lib/commands/update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const session_1 = require("../session");
const cli_1 = require("../cli");
exports.command = 'update';
exports.desc = 'Update Snips platform on your device';
exports.builder = {
    verbose: {
        demandOption: false,
        describe: 'Include remote logs',
        type: 'boolean',
        alias: 'v',
    },
};
exports.handler = (argv) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const verbose = JSON.parse(JSON.stringify(argv)).verbose;
    const ssh = new session_1.SSHService();
    yield ssh.connect()
        .catch(e => {
        cli_1.cli.stream.error(e);
        process.exit();
    });
    cli_1.cli.stream.println('Updating Snips Platform');
    yield ssh.updateSnips(output => {
        if (verbose) {
            cli_1.cli.stream.print(output);
        }
    })
        .catch(e => cli_1.cli.stream.error(e));
    cli_1.cli.stream.success('Finished updating the platform');
    ssh.disconnect();
    process.exit();
});
