// snips-sam/lib/commands/hass/install.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const chalk_1 = require("chalk");
const cli_1 = require("../../cli");
const ssh_1 = require("../../session/ssh");
exports.command = 'install';
exports.desc = 'Install home assistant on the device';
exports.builder = {
    verbose: {
        demandOption: false,
        describe: 'Include remote logs',
        type: 'boolean',
        alias: 'v',
    },
};
exports.handler = (argv) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const verbose = JSON.parse(JSON.stringify(argv)).verbose;
    const ssh = new ssh_1.SSHService();
    yield ssh.connect()
        .catch(e => {
        cli_1.cli.stream.error(e);
        process.exit();
    });
    cli_1.cli.stream.println(chalk_1.default.green('Installing Home Assistant with Python custom install'));
    yield ssh.installHomeAssistant(log => { if (verbose)
        cli_1.cli.stream.print(log); }, stepTitle => {
        if (verbose) {
            cli_1.cli.stream.println(chalk_1.default.green(stepTitle));
        }
        else {
            cli_1.cli.stream.success();
            cli_1.cli.stream.loading(stepTitle);
        }
    });
    ssh.disconnect();
    process.exit();
});
