// snips-sam/src/commands/hass/install.ts
import chalk from 'chalk'
import * as yargs from 'yargs'
import { cli } from '../../cli'
import { SSHService } from '../../session/ssh'

exports.command = 'install'
exports.desc = 'Install home assistant on the device'
exports.builder = {
    verbose: {
        demandOption: false,
        describe: 'Include remote logs',
        type: 'boolean',
        alias: 'v',
    },
}

exports.handler = async (argv: yargs.Argv) => {
    const verbose = JSON.parse(JSON.stringify(argv)).verbose

    const ssh = new SSHService()
    await ssh.connect()
    .catch(e => {
        cli.stream.error(e)
        process.exit()
    })

    cli.stream.println(chalk.green('Installing Home Assistant with Python custom install'))

    await ssh.installHomeAssistant(
        log => { if (verbose) cli.stream.print(log) },
        stepTitle => {
            if (verbose) {
                cli.stream.println(chalk.green(stepTitle))
            } else {
                cli.stream.success()
                cli.stream.loading(stepTitle)
            }
        })

    ssh.disconnect()
    process.exit()
}
